include/mqtt/message.hpp

```diff
@@ -529,46 +528,48 @@ class basic_publish_message {
         }
     }
 
-    template <typename Iterator>
-    basic_publish_message(Iterator b, Iterator e) {
-        if (b + 1 > e) throw remaining_length_error();
-        fixed_header_ = static_cast<std::uint8_t>(*b);
+    basic_publish_message(buffer buf) {
+        if (buf.empty())  throw remaining_length_error();
+        fixed_header_ = static_cast<std::uint8_t>(buf.front());
         auto qos = publish::get_qos(fixed_header_);
-        ++b;
+        buf = buf.substr(1);
 
-        if (b + 1 > e) throw remaining_length_error();
-        auto len_consumed = remaining_length(b, e);
+        if (buf.empty()) throw remaining_length_error();
+        auto len_consumed = remaining_length(buf.begin(), buf.end());
         remaining_length_ = std::get<0>(len_consumed);
-        auto consumed = static_cast<std::string::difference_type>(std::get<1>(len_consumed));
+        auto consumed = std::get<1>(len_consumed);
 
-        std::copy(b, b + consumed, std::back_inserter(remaining_length_buf_));
-        b += consumed;
+        std::copy(
+            buf.begin(),
+            buf.begin() + consumed,
+            std::back_inserter(remaining_length_buf_));
+        buf = buf.substr(consumed);
 
-        if (b + 2 > e) throw remaining_length_error();
-        std::copy(b, b + 2, std::back_inserter(topic_name_length_buf_));
-        auto topic_name_length = make_uint16_t(b, b + 2);
-        b += 2;
+        if (buf.size() < 2) throw remaining_length_error();
+        std::copy(buf.begin(), buf.begin() + 2, std::back_inserter(topic_name_length_buf_));
+        auto topic_name_length = make_uint16_t(buf.begin(), buf.begin() + 2);
+        buf = buf.substr(2);
 
-        if (b + topic_name_length > e) throw remaining_length_error();
-        utf8string_check(string_view(&*b, topic_name_length));
-        topic_name_ = as::buffer(&*b, topic_name_length);
-        b += topic_name_length;
+        if (buf.size() < topic_name_length) throw remaining_length_error();
+        utf8string_check(buf.substr(0, topic_name_length));
```

Comment:
Here, if buf has a shared_array_ptr, you're incrementing it's reference count for a function that we know ahead of time won't take ownership.

Better would be to do:

mqtt::string_view(buf.data(), buf.data()+topic_name_length)

Reply:
I updated the constructor parameter as follows:

```cpp
    basic_publish_message(mqtt::string_view buf) {
```